## TensorFlow
TensorFlow es una biblioteca de software de código abierto para computación numérica, que utiliza gráficos de flujo de 
datos. Los nodos en las gráficas representan operaciones matemáticas, mientras que los bordes de las gráficas 
representan las matrices de datos multidimensionales (tensores) comunicadas entre ellos.

## Keras 
Keras es una biblioteca de redes neuronales de código abierto escrita en Python. Es capaz de ejecutarse sobre 
TensorFlow, Microsoft Cognitive Toolkit o Theano. Está especialmente diseñada para posibilitar la experimentación en 
más o menos poco tiempo con redes de Aprendizaje Profundo. Sus fuertes se centran en ser amigable para el usuario, 
modular y extensible.

En 2017, el equipo de TensorFlow de Google decidió ofrecer soporte a Keras en la biblioteca de core de 
TensorFlow y desde entonces Keras es la herramienta principal de TensorFlow para el manejo de redes 
neuronales a nivel principiante gracias a que es muy fácil de manejar.

A continuación, se mostrara una pequeña guía de inicio rápido donde se mostraran los aspectos esenciales de la Keras 
y por lo tanto de TensorFlow 2.0.

In [1]:
import tensorflow as tf

Usualmente las paqueterías de Machine Learning o Deep Learning ya contienen **datasets** precargados con los cuales 
podemos empezar el estudio de estas. En este caso vamos a cargar el dataset de keras **mnist**

In [2]:
mnist = tf.keras.datasets.mnist

Cuando se estra trabajando con modelos de redes nueronales es común ver que la información contenida en un dataset 
se parte en 4 diferentes Arrays **x_train, y_train, x_test y y_test** los Arrays que tienen el nombre **train** se 
utilizan para entrenar la red neuronal y se deja una pequeña cantidad de datos para probar el performance después 
de entrenarla (por lo general se utiliza un 80% para el entrenamiento y un 20% para hacer el test).

In [3]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train/255.0, x_test/255.0 # se convierten los numero a punto flotante

11493376/11490434 [==============================] - 1s 0us/step


In [4]:
x_train

array([[[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       ...,

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0.

In [5]:
y_train

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [6]:
x_test

array([[[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       ...,

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0.

In [7]:
y_test

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

A la unidad de una red neuronal se le conoce como perceptrón, o lo que vendría siendo una sola neurona. Una red neuronal 
está compuesta por muchas neuronal que pasan la información de una capa a otra después de haber realizado el proceso 
interno de cada una de ellas. Las redes neuronales se dividen en capas, capa de entrada (input layer), capa escondida 
(hidden layer), y capa de salida (output layer). El calculo que una red neuronal realiza no es mas que una suma 
ponderada de los pesos que tiene cada uno de los valores de entrada. Si nosotros unimos muchas neuronas para formar 
una red neuronal sin aplicarle ninguna otra operación a los datos mas que la suma ponderada anteriormente mencionada 
la información resultante no nos seria de mucha utilidad, ya que la suma de funciones lineales siempre da como resultado
otra función lineal, esto realmente no da una representación correcta de los datos, por este motivo es que se aplican 
funciones de activación algunas de ellas son **relu** y **softmax**.

A continuación se construirá un modelo **tf.keras.sequential** apilando capas. 

In [8]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

Entrena y evalúa el modelo

Keras brinda muchas ventajas en comparación con otras paqueterías de Deep Learning, unas de ellas es, por ejemplo, 
que una vez entrenado el modelo te da la exactitud que tuvo al momento de probar la red con el set de datos divido 
para hacer el test

In [9]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test, y_test, verbose=2)

Epoch 1/5
1875/1875 [==============================] - 2s 953us/step - loss: 0.2972 - accuracy: 0.9134
Epoch 2/5
1875/1875 [==============================] - 2s 914us/step - loss: 0.1416 - accuracy: 0.9577
Epoch 3/5
1875/1875 [==============================] - 2s 948us/step - loss: 0.1079 - accuracy: 0.9671
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0887 - accuracy: 0.9729
Epoch 5/5
1875/1875 [==============================] - 2s 956us/step - loss: 0.0751 - accuracy: 0.9764
313/313 - 0s - loss: 0.0797 - accuracy: 0.9765


[0.07966028898954391, 0.9764999747276306]

El model de clasificacion de images fue entrenado y alcanzo una exactitud de ~98%